src/Components/Server/src/ComponentHub.cs

```diff
@@ -66,7 +66,15 @@ public override Task OnDisconnectedAsync(Exception exception)
             }
 
             CircuitHost = null;
-            return _circuitRegistry.DisconnectAsync(circuitHost, Context.ConnectionId);
+            if (exception != null)
+            {
+                return _circuitRegistry.DisconnectAsync(circuitHost, Context.ConnectionId);
+            }
+            else
+            {
+                Log.CircuitTerminatedGracefully(_logger, circuitHost.CircuitId);
+                return circuitHost.DisposeAsync().AsTask();
```

Comment:
Does this call the handler lifecycle methods?

Reply:
Yes